Updating browser history integration with the navigation Table of Contents (ToC) filter

RONAK-AI647 opened the pull request:
<!-- Please remove any unused sections -->

## Description
In account of the issue#209 , an issue arose where the navigation Table of Content (ToC) was not working properly . This PR talks about this issue only. Our  desired behavior in this updation was:

![image](https://github.com/user-attachments/assets/a24fc3e4-c858-4a57-aedb-0dd2e16fa43d)


#### Issue addressed
#213 
Addresses #*PR# HERE*

### Before/after screenshots 
The first URL

![image](https://github.com/user-attachments/assets/95a02237-7e79-4fd2-94d3-3afd20dc1fe7)

When button is typed in ToC

![Screenshot 2024-12-01 005300](https://github.com/user-attachments/assets/2556a2f1-2287-4d7e-8eae-020a165de5c0)

When button is pressed

![Screenshot 2024-12-01 005415](https://github.com/user-attachments/assets/6d457ed0-f190-49d9-9d90-e74ecc57762d)

FIrst back press

![Screenshot 2024-12-01 005503](https://github.com/user-attachments/assets/039b0041-95ab-4971-9990-ebb56cd1feb4)

Second back press

![image](https://github.com/user-attachments/assets/f3c0b5eb-e65b-4dc7-bca4-55bd87b6cbf0)



## Changelog
<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG START -->

<!--
  - Fill in the changelog item(s) below. If there are more groups of closely
    related changes, prepare more changelog items for each one of them.
    At a minimum, always separete non-breaking changes from breaking changes.
  - This needs to be pasted to CHANGELOG.md before merging a PR.
  - See changelog guidelines https://www.notion.so/learningequality/DRAFT-Changelog-Guidelines-106b6ebbdeda4ba5b3b3e7c490c5a4fe and existing
    items in CHANGELOG.md as examples
 -->

  - **Description:** Improvement of the Table of Contents filter to work properly with browsers' command.
  - **Products impact:** Updated API
  - **Addresses:** Issue#213 
  - **Components:** no
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** Improves KDS.
<!-- [DO NOT REMOVE-USED BY GH ACTION] CHANGELOG END -->

## Steps to test

1. Open KDS on local server .
2. The modification done in index.vue under sidenav section.
3. .........

## (optional) Implementation notes

### At a high level, how did you implement this?
<!-- Briefly describe how this works -->

### Does this introduce any tech-debt items?
<!-- List anything that will need to be addressed later -->

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## Comments
The watch and mount section in docs\common\DocsPageTemplate\SideNav\index.vue is modified.

RONAK-AI647 committed:
pushing to browser history

RONAK-AI647 committed:
update index.vue

RONAK-AI647 committed:
javascript check

RONAK-AI647 committed:
Restoring filter state when a user navigates back

RONAK-AI647 committed:
Update vue.index

RONAK-AI647 committed:
javascript test

RONAK-AI647 committed:
event to events

RONAK-AI647 committed:
revert the commit

RONAK-AI647 committed:
Update index.vue

RONAK-AI647 committed:
from search to filter

RONAK-AI647 committed:
The URL perfecting

RONAK-AI647 committed:
Revert "The URL perfecting"

This reverts commit 7ee60495104dd3e7e7b4a6010e2225d53e983d55.

RONAK-AI647 committed:
Updating watch for URL fix

RONAK-AI647 committed:
Spacing and linting

RONAK-AI647 committed:
lint-test

RONAK-AI647 commented:
I wonder if the maintainers could pass the linter test for me , as the linter does not fix the issues in windows .

MisRob commented:
Thank you @RONAK-AI647! I will invite my colleagues for review.

akolson commented:
Hi @RONAK-AI647 I am not sure whether this is within the scope if this task but I did pick up the issue below; Directly accessing the for example `/?filter=kcard` via url bar doesn't trigger the filtering of the TOC. @MisRob @AlexVelezLl any thoughts?

https://github.com/user-attachments/assets/f65665f9-341a-492c-81cb-ab5bde8900a5

RONAK-AI647 commented:
@akolson I see it , but it works when you paste the extension **"/kcard?filter=kcard"** rather than     **" /?filter=kcard "**.  because the URL was suppposed to be this , according to the box below.

![image](https://github.com/user-attachments/assets/bc4491ec-8806-4540-a389-62353a3c4220)

If you say that   pasting this URL extension  **" /?filter=kcard "** does not enter the button's name in Toc filter , yes it does not , I see that .

akolson commented:
@RONAK-AI647 unfortunately I can also replicate this 
> but it works when you paste the extension "/kcard?filter=kcard" rather than " /?filter=kcard ". because the URL was suppposed to be this , according to the box below.

You can check here on your [pr deployment](https://deploy-preview-847--kolibri-design-system.netlify.app/)

https://github.com/user-attachments/assets/2a767896-d58b-47c4-bfdf-86ae9c341832

RONAK-AI647 commented:
@akolson , this could be beyond the scope of this task , i can be wrong , if you all confirm it as an emerging issue , i will work on that as well.

RONAK-AI647 commented:
https://github.com/user-attachments/assets/aafb1541-a62f-4c02-891c-f7ed2833f709

I see ,this was previously with the KDS , so i guarantee my new codes did not generate this error 😅. @akolson @MisRob

AlexVelezLl commented:
Hi @RONAK-AI647!! Thanks for working on this issue :). Yes, this issue is in the scope of this PR. Previous to this PR we didnt had a `filter` query in the route, so any bug we have with this route query is because we have introduced some error with this PR.
 
A couple of things that can be originating this error:
1. We need to remove the previous handlers that stored this filter value in the sessionStorage, because we are not relying in the sessionStorage anymore to handle this value, but on the route query, and we need to have a single source of truth: 
  * This line: https://github.com/RONAK-AI647/kolibri-design-system/blob/ac409f6b908ff90f9487bba17bb784c3aa11867e/docs/common/DocsPageTemplate/SideNav/index.vue#L80.
  * And these lines: https://github.com/RONAK-AI647/kolibri-design-system/blob/ac409f6b908ff90f9487bba17bb784c3aa11867e/docs/common/DocsPageTemplate/SideNav/index.vue#L93-L96
2. After we do this :arrow_up: the error reported by @akolson becomes more obvious, when we reload the page we lose the filter query:

  https://github.com/user-attachments/assets/48f4765c-db62-441f-b48e-c044fbe071bc

3. We have missed an instruction described in #213:
  > Persist [trailing debounced changes](https://lodash.com/docs/#debounce) (wait time ~2s) to the filter as query string changes and push them onto the browser history stack.

  We need to _push_ the filter as query debounced by 2s (we can play with the wait time if needed), and we need to read these chages so when we reload the page or go back we have this value set. After we _push_ the query, we dont need to do any of this https://github.com/RONAK-AI647/kolibri-design-system/blob/ac409f6b908ff90f9487bba17bb784c3aa11867e/docs/common/DocsPageTemplate/SideNav/index.vue#L99-L106 because the browser history is the one that will be in charge of managing the state changes, so we dont need to listen to the pop.

Hope these instructions help. Let me know if you have any questions :).

RONAK-AI647 committed:
Removing previous handlers

RONAK-AI647 commented:
Thank You @AlexVelezLl for the guidance .🥰
Thank You @akolson for reviews about the PR, I am working on it , Soon post the outcome.

RONAK-AI647 committed:
Pushing filter as query

RONAK-AI647 commented:
https://github.com/user-attachments/assets/f27ff59d-9eb5-4f51-8528-551ed54ddacc


**Not loosing filter query when page reloads .** @AlexVelezLl

RONAK-AI647 committed:
Adding debounceUpdate

RONAK-AI647 committed:
Adding debounce under mounted()

RONAK-AI647 commented:
Hey members , take a look please. @AlexVelezLl @akolson 

1) The filter Toc and browser's forward and backward button.

https://github.com/user-attachments/assets/027b06fd-e21c-4e07-a497-931f4c8bbb18

2) Not loosing filter query when page reloads

https://github.com/user-attachments/assets/6b3e296e-2555-4c0f-a601-3ba834040d15

3) Directly accessing from the URL bar.

https://github.com/user-attachments/assets/cc318ae8-c6f0-4dce-b574-1cbead5a1429

AlexVelezLl reviewed (commented):
Thank you @RONAK-AI647!! This is looking great! We are almost there, the only thing I have noticed that is missing is that after we click on a component, we dont preserve the filter query and it gets deleted, this is the only behaviour missing from the table that was presented in the issue:

| URL | action |
|--|--|
| `/` | first URL|
| `/?filter=button` | after enter 'button' into ToC filter |
| `/kbutton?filter=button` | after clicking `KButton` entry |
| `/?filter=button` | browser back |
| `/` | browser back |

Its the 3rd action what is still missing. Apart from that everything looks good!

AlexVelezLl commented:
Also, can you please rebase your branch on top of the latest KDS develop? We recently updated the KDS nodejs version to v18 and vue to v2.7, and we need to do the rebase for the actions to run successfully :). There are also new linting rules, so after rebasing, please run `yarn lint-fix` again to comply with the new rules. Thanks!

RONAK-AI647 committed:
Preserving the query when the button is clicked

RONAK-AI647 commented:
@AlexVelezLl   I have considered the 3rd step and now its preserved. You can take a look .


https://github.com/user-attachments/assets/1ab60b98-e708-4eaf-81fa-58922f3e7d26

RONAK-AI647 commented:
@MisRob @AlexVelezLl , I face a problem here , i rebased the pr , the lint fix yet shows list of errors in the terminal. Here the common cicd checks also are failing , any idea , what is wrong .
1) I updated the node to version 18. and vue to vue 2.


https://github.com/user-attachments/assets/cb661f71-201d-4160-b049-30d4482c1299

AlexVelezLl commented:
Hey @RONAK-AI647! Did you run the `yarn install` again after rebasing?

RONAK-AI647 commented:
> Hey @RONAK-AI647! Did you run the `yarn install` again after rebasing?

yes, I did yarn install. , yet the yarn lint-fix does not work I don't understand why these tests fail.
![image](https://github.com/user-attachments/assets/c4f4ca74-eda4-48ae-bab1-fa7163e1451d)

do i open a new PR???

AlexVelezLl commented:
Hey @RONAK-AI647! Its possible that the error you are facing is because of a specific setting in your computer. Can you explore the answers described in this thread: https://stackoverflow.com/questions/8965606/node-and-error-emfile-too-many-open-files. I was able to run `yarn lint-fix` with your branch, so its probably something to do with the OS.

I will further investigate the failing netlify actions, these shouldnt be failing.

AlexVelezLl commented on docs/common/DocsPageTemplate/SideNav/index.vue:
```diff
@@ -80,22 +82,62 @@
+    }
+
+  // Modify navigation links to preserve the filter query
+  this.$nextTick(() => {
```
I would suggest better update [NavLink](https://github.com/AlexVelezLl/kolibri-design-system/blob/e9aa0abdff8c659568cb5dd36a38a3dc847cec2f/docs/common/DocsPageTemplate/SideNav/NavLink.vue) and [DocsLibraryLink](https://github.com/AlexVelezLl/kolibri-design-system/blob/e9aa0abdff8c659568cb5dd36a38a3dc847cec2f/docs/common/DocsPageTemplate/SideNav/NavLink.vue) to pass the query params, so we have dont have to add an event listener to all links of the page :)

RONAK-AI647 committed:
Lint -fix (ubuntu version)

RONAK-AI647 commented:
@AlexVelezLl  i got the linting problem  and the conflicts with rest of the checks will be dealt if i open a new PR . The points you mentioned , i will keep in mind and soon i will a new PR with everything resolved. Thank you

AlexVelezLl committed:
Force deploy

AlexVelezLl commented:
I managed to get it working @RONAK-AI647! Here is the deployed site https://deploy-preview-847--kolibri-design-system.netlify.app/. You wont need to open a different PR :open_hands:.

RONAK-AI647 commented:
> I managed to get it working @RONAK-AI647! Here is the deployed site https://deploy-preview-847--kolibri-design-system.netlify.app/. You wont need to open a different PR 👐.

super Thank You !!!!!!

RONAK-AI647 committed:
Updating Navlink.vue for preserving query param

RONAK-AI647 committed:
Update DocsLibraryLink.vue

RONAK-AI647 committed:
yarn lint-fix

RONAK-AI647 committed:
Merge branch 'navigating-browser-history-issue' of https://github.com/RONAK-AI647/kolibri-design-system into RONAK-A1647#213-Improve-browser-history-integration-with-the-navigation-Table-of-Contents-(ToC)-filter

RONAK-AI647 commented:
@AlexVelezLl  I updated the navlink.vue and DocsLibraryLink.vue as you said , take a look please.

RONAK-AI647 committed:
Removing previous handler in index.vue

AlexVelezLl commented on docs/common/DocsPageTemplate/SideNav/index.vue:
```diff
@@ -77,22 +79,54 @@
+              query: { ...this.$route.query, filter: newValue },
+            });
+          }
+          this.debouncedUpdateQuery(newValue);
```
As you took another approach with the popstate listener, lets just remove this `debouncedUpdateQuery` call. As there we just do the same logic of pushing the query.

```suggestion
```

AlexVelezLl commented on docs/common/DocsPageTemplate/SideNav/index.vue:
```diff
@@ -77,22 +79,54 @@
-          window.sessionStorage.setItem('nav-filter', newValue);
+          //Clear the filter query when filtertext is empty
+          if (!newValue) {
+            this.$router.replace({ path: this.$route.path, query: {} });
```
I think we can just push the new route, instead of replacing it.

AlexVelezLl commented on docs/common/DocsPageTemplate/SideNav/index.vue:
```diff
@@ -77,22 +79,54 @@
+          }
+        });
+        // Create debounced function for updating the query
+        this.debouncedUpdateQuery = debounce(newValue => {
+          if (newValue) {
+            this.$router.push({
+              path: this.$route.path,
+              query: { ...this.$route.query, filter: newValue },
+            });
+          } else {
+            this.$router.push({ path: this.$route.path, query: {} });
+          }
+        }, 2000); // 2-second debounce delay
```
As we wont be using this debouncedUpdateQuery anymore, we can just remove these lines.
```suggestion
```

AlexVelezLl commented on docs/common/DocsPageTemplate/SideNav/index.vue:
```diff
@@ -51,6 +52,7 @@
       return {
         filterText: '',
         loaded: false,
+        debounceUpdateQuery: null,
```
And dont forget to remove this data variable too :)

AlexVelezLl replied:
Hi @RONAK-AI647! We are not using this state anymore, we can remove it too :)

RONAK-AI647 replied:
> Hi @RONAK-AI647! We are not using this state anymore, we can remove it too :)

ahh I saw that , yet i missed , my bad !!!!!!

AlexVelezLl reviewed (commented):
Hey @RONAK-AI647! This looks good to me! We are almost there, just suggested to remove the `debounceUpdateQuery` method as it seems is not needed anymore, since the intention of the debounce has been addressed in another way. After this, we can merge the PR. Thank you!

RONAK-AI647 committed:
Removing debouncedUpdateQuery

RONAK-AI647 commented:
@AlexVelezLl , made the changes.

RONAK-AI647 committed:
Update index.vue

akolson reviewed (commented):
LGTM! both changes and manual QA checkout. @AlexVelezLl  any other thoughts? Please proceed to give final approval and merge once everything is clear in your side concerning your feedback. Thanks

AlexVelezLl reviewed (approved):
LGTM! Thanks for all your work @RONAK-AI647!
